### PR TITLE
fix: load moltbook agent tokens from env

### LINKED ISSUE #4609: Hardcoded Moltbook bearer tokens in solver script
## Summary
`scripts/moltbook_solver.py` committed six Moltbook bearer-token shaped agent credentials directly in the `AGENTS` registry. The same module uses those keys as `Authorization: Bearer ...` credentials for `https://www.moltbook.com/api/v1/posts` and `/api/v1/verify`.

## Impact
Anyone with repository read access can recover the embedded agent credentials and use them to post or verify as those Moltbook agents until the keys are rotated. Because the solver includes agent rotation and auto-verification, leaked keys can be abused for unauthorized posting and account suspension risk.

## Fix
Move each agent credential to a per-agent environment variable, only mark agents available when their key is configured, and update tests to avoid committing secret-shaped Moltbook tokens. Existing exposed Moltbook credentials should be rotated out of band.

## Validation
- `rg -n "moltbook_sk_[A-Za-z0-9_-]+" scripts node tests tools integrations --glob "!**/get-pip.py" --glob "!**/package-lock.json"` -> no matches
- `python3 -m py_compile scripts/moltbook_solver.py scripts/tests/test_moltbook_solver.py` -> passed
- `uv run --no-project --with pytest --with requests python -m pytest scripts/tests/test_moltbook_solver.py -q` -> 23 passed
- `python3 tools/bcos_spdx_check.py --base-ref origin/main` -> OK

### PR DESCRIPTION
**Pull request by ethever**
## Summary
- Fixes #4609
- Removes committed Moltbook agent bearer-token values from `scripts/moltbook_solver.py`
- Stores only per-agent environment variable names in the agent registry
- Excludes agents from rotation unless their token environment variable is configured
- Updates solver tests so they do not rely on secret-shaped test values

## Validation
- `rg -n "moltbook_sk_[A-Za-z0-9_-]+" scripts node tests tools integrations --glob "!**/get-pip.py" --glob "!**/package-lock.json"` -> no matches
- `python3 -m py_compile scripts/moltbook_solver.py scripts/tests/test_moltbook_solver.py`
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 uv run --no-project --with pytest --with requests python -m pytest scripts/tests/test_moltbook_solver.py -q` -> 24 passed
- `git diff --check origin/main...HEAD`
- `python3 tools/bcos_spdx_check.py --base-ref origin/main`

## Bounty
Wallet: b3a58f80a97bae5e2b438894aa85600cb0c066RTC

### CODE REVIEW

**Comment by ethever** [comment]
Checklist for #4609:

- No Moltbook bearer-token shaped values remain in the searched source/test paths.
- Agent tokens now come from per-agent env vars such as `MOLTBOOK_AGENT_SOPHIA_KEY`.
- `get_available_agents()` returns only agents with configured keys.
- Focused tests and BCOS SPDX checks passed.

Bounty wallet: `b3a58f80a97bae5e2b438894aa85600cb0c066RTC`

**Comment by 508704820** [comment]
## Code Review: Load moltbook agent tokens from environment variables

### Summary
Removes **6 hardcoded API keys** (`moltbook_sk_*`) from source code and loads them from environment variables instead.

### Critical Finding 🚨
The original code contained **6 hardcoded secret keys** (CWE-798):
```
moltbook_sk_nuTK8FxFHuUtknLGrXUJKxcgBsTJ0zP7   (sophia)
moltbook_sk_mACTltXU55x6s1mYqDuWkeEcuDQ9feMB   (boris)
moltbook_sk_yWpLPPIp1MxWAlbgiCEdamHodyClGg08   (janitor)
moltbook_sk_CJgvb5ecA9ZnutcmmaFy2Scm_X4SQgcz   (bottube)
moltbook_sk_-zuaZPUGMVoC_tdQJA-YaLVlj-VnUMdw   (msgoogletoggle)
moltbook_sk_BeO3rZoBKuleNwSX3sZeBNQRYhOBK436   (oneo)
```

These are now **publicly visible** in the Git history. All 6 keys should be **rotated immediately** — simply removing them from the code is not enough since they're already in the commit history.

### Positive ✅
1. **Removed all 6 hardcoded keys** — proper security practice
2. **`key_env` pattern** — each agent reads from a dedicated env var
3. **Agent availability check** — `get_agent_key(a)` returns None for unconfigured agents, excluding them from the available list
4. **Graceful degradation** — agents without configured keys are silently skipped

### Note
All 6 `moltbook_sk_*` keys should be **rotated/revoked** on the Moltbook service since they were exposed in the public repo.

**LGTM** ✅ — critical hardcoded credential removal.

**Review quality: Security-focused review** (25 RTC)

File: scripts/moltbook_solver.py [hunk 1]
```diff
@@ -124,12 +142,18 @@ def get_available_agents() -> List[str]:
     # Preference order: msgoogletoggle first (it's our best solver host),
     # then sophia, boris, janitor, bottube, oneo
     preferred = ["msgoogletoggle", "sophia", "boris", "janitor", "bottube", "oneo"]
-    return [a for a in preferred if a in AGENTS and a not in suspended]
+    return [
+        a for a in preferred
+        if a in AGENTS and a not in suspended and get_agent_key(a)
```
Comment (dicnunz) [inline scripts/moltbook_solver.py:147]: Because this predicate now excludes agents with unset or blank `MOLTBOOK_AGENT_*_KEY` values, `get_available_agents()` no longer means only "not suspended". `post_with_rotation()` and `self_test()` still label an empty list as `ALL agents suspended` / `SUSPENDED`, so a deployment with missing env vars gets a false suspension diagnosis. Non-blocking for the secret-removal fix, but consider wording this as unavailable/unconfigured or splitting configured-vs-suspended status.

**Review by dicnunz — APPROVED** [review]
Approved after focused review of PR #4928 at head e23e8e8e97ac7ca7092ba932d06b8a47b01fde15.

What I checked:
- The committed Moltbook bearer-token literals are removed from the agent registry and replaced with per-agent environment variable names.
- `get_agent_key()` resolves credentials at call time and trims blank env values to `None`.
- `get_available_agents()` filters suspended agents and unconfigured agents before `post_with_rotation()` can attempt a bearer request, so empty/missing credentials are not sent.
- The new tests cover the no-configured-key case, env-backed key lookup, missing-env behavior, required registry fields, and absence of the Moltbook token prefix in `scripts/moltbook_solver.py`.
- I compared this against the older same-issue PR #4610; this PR uses the `MOLTBOOK_AGENT_*_KEY` naming and adds stronger no-token/no-config regression coverage.

Validation run locally against `origin/main...HEAD`:
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 uv run --no-project --with pytest --with requests python -m pytest scripts/tests/test_moltbook_solver.py -q` -> 24 passed
- `python3 -m py_compile scripts/moltbook_solver.py scripts/tests/test_moltbook_solver.py` -> passed
- `git diff --check origin/main...HEAD` -> passed
- `python3 tools/bcos_spdx_check.py --base-ref origin/main` -> OK
- `rg -n "moltbook_sk_[A-Za-z0-9_-]+" scripts node tests tools integrations --glob "!**/get-pip.py" --glob "!**/package-lock.json"` -> no matches
- Local env probe confirmed no configured keys returns `[]`, configured `msgoogletoggle`/`sophia` preserve preferred order, and token values are stripped.

No blocking findings. One non-blocking operational wording note inline: after this patch, an empty availability list may mean missing env configuration, not only suspension.

**Review by saim256 — APPROVED** [review]
Reviewed current head `e23e8e8e97ac7ca7092ba932d06b8a47b01fde15`.

The committed Moltbook bearer-token values are removed from `scripts/moltbook_solver.py` and replaced with per-agent environment variable names. `get_agent_key()` resolves and strips those env values at runtime, and `get_available_agents()` now excludes agents without configured keys, so rotation will not attempt a bearer request with missing credentials.

The tests cover no configured keys, env-backed key lookup, missing-env behavior, required registry fields, and absence of the Moltbook secret prefix in the solver source. Operationally, the six exposed historical keys should still be rotated/revoked because they remain public in Git history.

Validation performed locally:
- `python -m pytest scripts\tests\test_moltbook_solver.py -q` -> 24 passed
- `python -m py_compile scripts\moltbook_solver.py scripts\tests\test_moltbook_solver.py` -> passed
- `rg -n "moltbook_sk_[A-Za-z0-9_-]+" scripts node tests tools integrations --glob "!**/get-pip.py" --glob "!**/package-lock.json"` -> no matches
- `python -m ruff check scripts\moltbook_solver.py scripts\tests\test_moltbook_solver.py --select E9,F821,F811 --output-format=concise` -> passed
- `git diff --check origin/main...HEAD` -> passed
- `python tools\bcos_spdx_check.py --base-ref origin/main` -> OK

No live Moltbook API call, production token, wallet, or destructive testing was used.

**Review by TJCurnutte — APPROVED** [review]
Reviewed the Moltbook agent-token cleanup.

Validation run on the PR head:
- `python3 -m py_compile scripts/moltbook_solver.py scripts/tests/test_moltbook_solver.py` -> passed
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 uv run --no-project --with pytest --with requests python -m pytest scripts/tests/test_moltbook_solver.py -q` -> passed
- Local secret scan across the changed solver/test files for `moltbook_sk_...` token-shaped values -> no matches
- `git diff --check origin/main...HEAD` -> passed
- `python3 tools/bcos_spdx_check.py --base-ref origin/main` -> OK

The registry now stores environment-variable names instead of committed bearer-token values, unavailable agents are excluded from rotation, and the tests cover both missing and configured env-key behavior.

**Review by loganoe — APPROVED** [review]
Approved PR #4928 on current head `e23e8e8e97ac7ca7092ba932d06b8a47b01fde15`.

The committed `moltbook_sk_...` bearer-token literals are removed from `scripts/moltbook_solver.py` and replaced with per-agent environment variable names. `get_agent_key()` resolves trimmed env values at runtime, and `get_available_agents()` filters out agents without configured keys before rotation can try to post with missing credentials. The tests cover no configured keys, env-backed lookup, missing-env behavior, registry fields, and absence of the Moltbook secret prefix in the solver source.

Validation I ran:
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 uv run --no-project --with pytest --with requests python -m pytest scripts/tests/test_moltbook_solver.py -q` -> 24 passed
- `python3 -m py_compile scripts/moltbook_solver.py scripts/tests/test_moltbook_solver.py` -> passed
- `rg -n "moltbook_sk_[A-Za-z0-9_-]+" scripts node tests tools integrations --glob "!**/get-pip.py" --glob "!**/package-lock.json"` -> no matches
- `git diff --check origin/main...HEAD` -> passed
- `uv run --no-project python tools/bcos_spdx_check.py --base-ref origin/main` -> OK
- `uv run --no-project --with ruff ruff check scripts/moltbook_solver.py scripts/tests/test_moltbook_solver.py --select E9,F821,F811,F841` reports an existing F841 in `scripts/moltbook_solver.py`; I confirmed the same finding is present on `origin/main`, so it is not introduced by this PR.

Operational note: the exposed historical Moltbook keys should still be rotated/revoked, because removing them from the working tree does not remove them from Git history.

No live Moltbook API call, production token, wallet, or destructive testing was used.

Bounty #73 payout wallet if this review is eligible: `6Da5nELroja5ngTwYZuofFur5V7gZCLvKVRX7iUahwz2`

**Review by himanalot — APPROVED** [review]
Reviewed the Moltbook token handling fix. The committed bearer values are replaced with per-agent environment variable names, `get_agent_key()` now resolves tokens from the environment, and unconfigured agents are excluded from rotation rather than using a fallback secret. The updated tests cover missing env keys, env lookup, registry shape, and absence of committed token prefixes. I do not see a blocking issue in this change.

**Comment by Scottcjn** [comment]
Wave-5 cluster pay sent (4 RTC/PR). Please rebase against current main to clear conflicts — multiple wave-5 PRs touch overlapping files.

**Comment by Scottcjn** [comment]
Wave-5 cluster pay already issued (you received 88 RTC earlier today). This PR is conflict-blocked against current main — overlapping changes from concurrent wave-5 merges.

To land: rebase your branch on current main, resolve conflicts (mostly in `node/rustchain_v2_integrated_v2.2.1_rip200.py` and a few other shared files), and force-push. We'll auto-merge once it's MERGEABLE+CLEAN.

No additional pay needed — cluster rate locked the value. Just need the code in main.

If you don't want to rebase, close the PR and the wave-5 pay still stands. Your call.

**Review by jaxint — APPROVED** [review]
## PR Review Summary

**LGTM** ✅ — Solid contribution to the RustChain ecosystem.

### Code Quality
- Implementation follows project conventions
- Security considerations adequately addressed  
- Error handling appears robust

### Testing
- Test coverage is appropriate for the changes
- Edge cases covered

**Approved.** Ready to merge. 🚀

**Review by jaxint — APPROVED** [review]
## PR Review Summary

**LGTM** ✅ — Solid contribution to the RustChain ecosystem.

### Code Quality
- Implementation follows project conventions
- Security considerations adequately addressed  
- Error handling appears robust

### Testing
- Test coverage is appropriate for the changes
- Edge cases covered

### Documentation
- Code is readable and self-documenting
- Comments where needed

**Approved.** Ready to merge. 🚀

**Review by jaxint — APPROVED** [review]
LGTM! Thanks for contributing. Approved.

**Comment by Scottcjn** [comment]
Heads up — codex audit gave this PR a thumbs-up (high-tier security work: *Removes committed Moltbook tokens; rebase plus secret rotation needed.*), but it currently conflicts with `main` after the cerredz cluster merged. Could you rebase against the current `main` tip? Once green, queued for 50 RTC payout.

**Review by HCIE2054 — COMMENTED** [review]
Reviewed PR 4928. Standard review.

**Review by HCIE2054 — APPROVED** [review]
LGTM! Thanks for contributing to RustChain!

**Review by HCIE2054 — APPROVED** [review]
LGTM! Thanks for contributing to RustChain!

**Review by HCIE2054 — APPROVED** [review]
LGTM! Great contribution to the RustChain ecosystem. Thanks for keeping the codebase clean and well-tested. Approved ✅

**Review by HCIE2054 — APPROVED** [review]

**Review by 508704820 — COMMENTED** [review]
Load moltbook agent tokens from env. Verify: (1) Tokens are never hardcoded or committed. (2) Missing env var = fail closed, not empty token. (3) Token rotation is supported. - Xeophon (security review, secrets)

**Review by HCIE2054 — APPROVED** [review]
LGTM!

**Review by HCIE2054 — APPROVED** [review]
LGTM!

**Comment by Scottcjn** [comment]
Closing as **stale branch — would cause destructive deletions if merged.**

Your branch was filed roughly **966 commits behind current main**. Since then, many overlapping fixes from other contributors have landed via parallel PRs. GitHub's `mergeable=clean` indicator only catches text-level conflicts — it doesn't detect "this branch is so old that merging would silently delete code that landed after it was filed."

## Bounty credit acknowledged

If your fix addressed a real bug, the canonical version has very likely already shipped via a parallel contributor's PR over the past two weeks. Specific cases covered by today's audit:

- **Age-scoring (`2025` → `current_year` fixes)** — credited in synthesis PR #5660 (Co-Authored-By: saim256, slashdevcorpse, galpetame, ethever, junn-dev)
- **Auth strengthening (admin-key gating, signed wallet/OTC actions)** — landed via earlier merges this week
- **Input validation (`fix: validate X`, `fix: reject X`)** — landed via merged contributor first-posters
- **CWE-209 info-disclosure hardening (`hide X exception details`)** — landed via @hungle123-dev first-poster PRs
- **CONTRIBUTING.md per-repo** — closed bounty #1605 with first-poster awards on 2026-05-18

## If you want fresh review

Rebase against current main and verify your diff shows roughly the size of the changes you originally made:

```
git fetch upstream main
git rebase upstream/main
git diff main..HEAD --shortstat
```

If the deletion count is significantly higher than what you added, the branch is still picking up stale assumptions — recreate from a fresh main.

Thanks for the contribution work this week.
